src/bmffimage.cpp

```diff
@@ -0,0 +1,521 @@
+            // hunt for "II" or "MM"
+            long  eof  = 0xffffffff; // impossible value for punt
+            long  punt = eof;
+            for ( long i = 0 ; i < exif.size_ -8 && punt==eof ; i+=2) {
+                if ( exif.pData_[i] == exif.pData_[i+1] )
+                    if ( exif.pData_[i] == 'I' || exif.pData_[i] == 'M' )
+                        punt = i;
+            }
+            if ( punt != eof ) {
+                Internal::TiffParserWorker::decode(exifData(), iptcData(), xmpData(),
+                  exif.pData_+punt, exif.size_-punt, root_tag,
+                  Internal::TiffMapping::findDecoder);
+            }
+        }
+        io_->seek(restore,BasicIo::beg);
+    }
+
+    void BmffImage::parseTiff(uint32_t root_tag, uint32_t length)
+    {
+        if (length > 8) {
+            DataBuf data(length - 8);
+            long bufRead = io_->read(data.pData_, data.size_);
+
+            if (io_->error())
+                throw Error(kerFailedToReadImageData);
+            if (bufRead != data.size_)
+                throw Error(kerInputDataReadFailed);
+
+            Internal::TiffParserWorker::decode(exifData(), iptcData(), xmpData(),
+                                               data.pData_, data.size_, root_tag,
+                                               Internal::TiffMapping::findDecoder);
+        }
+    }
+
+    void BmffImage::setComment(const std::string& /*comment*/)
+    {
+        // Todo: implement me!
+        throw(Error(kerInvalidSettingForImage, "Image comment", "BMFF"));
+    }  // BmffImage::setComment
+
+    void BmffImage::readMetadata()
+    {
+        if (io_->open() != 0) {
+            throw Error(kerDataSourceOpenFailed, io_->path(), strError());
+        }
+        IoCloser closer(*io_);
+        // Ensure that this is the correct image type
+        if (!isBmffType(*io_, false)) {
+            if (io_->error() || io_->eof())
+                throw Error(kerFailedToReadImageData);
+            throw Error(kerNotAnImage, "BMFF");
+        }
+        ilocs_.clear();
+        visits_.clear();
+        visits_max_ = io_->size() / 16;
+
+        unknownID_ = 0xffff;
+        exifID_ = unknownID_;
+
+        long address = 0;
+        while (address < (long)io_->size()) {
+            io_->seek(address, BasicIo::beg);
+            address = boxHandler();
+        }
+    }  // BmffImage::readMetadata
+
+    void BmffImage::printStructure(std::ostream& out, PrintStructureOption option, int depth)
+    {
+        if (io_->open() != 0)
+            throw Error(kerDataSourceOpenFailed, io_->path(), strError());
+
+        // Ensure that this is the correct image type
+        if (!isBmffType(*io_, false)) {
+            if (io_->error() || io_->eof())
+                throw Error(kerFailedToReadImageData);
+            throw Error(kerNotAnImage);
+        }
+        UNUSED(out);
+        UNUSED(option);
+        UNUSED(depth);
+    }  // BmffImage::printStructure
```

Comment:
What's the plan for this function? Shouldn't we implement this too?

Reply:
I added the printStructure() code in 2013 when Tuan as a GSoC Student with us and developed `class RemoteIo`.  He asked the question _"Where is the metadata in the file?"_ and my answer was _"only Andreas knows!"_.  Tuan investigated a couple of formats: JPEG and PNG and wrote articles for dev.exiv2.org

The exiv2 command options `-pS`, `-pR` reveal the internal structure of the file and `-pX` and `-pC` reveal the "raw" XMP and ICC profiles.

If we had more time, I'd say "implement it" and we could do that by adding an option to boxHandler().  On the other hand, tvisitor works well.  The default implementation of image::printStructure() is to throw.  Let's do that.  For example `gif` files are supported in Exiv2:

```
986 rmills@rmillsm1:~/gnu/exiv2/team/book $ exiv2 files/GIF.gif 
File name       : files/GIF.gif
File size       : 2321 Bytes
MIME type       : image/gif
Image size      : 8 x 8
files/GIF.gif: No Exif data found in the file
987 rmills@rmillsm1:~/gnu/exiv2/team/book $ 
```

However gifimage.cpp uses the default image::printStructure() which throws when you use an option such as `-pS`

```bash
985 rmills@rmillsm1:~/gnu/exiv2/team/book $ exiv2 -pS files/GIF.gif 
Exiv2 exception in print action for file files/GIF.gif:
Image type files/GIF.gif is not supported
986 rmills@rmillsm1:~/gnu/exiv2/team/book $ 
```

To explain a little more.  Here's the output from `-pS` on a JPEG:
```bash
987 rmills@rmillsm1:~/gnu/exiv2/team/book $ exiv2 -pS files/Stonehenge.jpg 
STRUCTURE OF JPEG FILE: files/Stonehenge.jpg
 address | marker       |  length | data
       0 | 0xffd8 SOI  
       2 | 0xffe1 APP1  |   15272 | Exif..II*......................
   15276 | 0xffe1 APP1  |    2786 | http://ns.adobe.com/xap/1.0/.<?x
   18064 | 0xffed APP13 |      96 | Photoshop 3.0.8BIM.......'.....
   18162 | 0xffe2 APP2  |    4094 | MPF.II*...............0100.....
   22258 | 0xffdb DQT   |     132 
   22392 | 0xffc0 SOF0  |      17 
   22411 | 0xffc4 DHT   |     418 
   22831 | 0xffda SOS  
988 rmills@rmillsm1:~/gnu/exiv2/team/book $
```

And from tvisitor:
```bash
988 rmills@rmillsm1:~/gnu/exiv2/team/book $ tvisitor files/Stonehenge.jpg 
STRUCTURE OF JPEG FILE (II): files/Stonehenge.jpg
 address | marker       |  length | signature
       0 | 0xffd8 SOI  
       2 | 0xffe1 APP1  |   15272 | Exif__II*_.___._..._.___.___..._.___.___
   15276 | 0xffe1 APP1  |    2786 | http://ns.adobe.com/xap/1.0/_<?xpacket b
   18064 | 0xffed APP13 |      96 | Photoshop 3.0_8BIM.._____'..__._...Z_..%
   18162 | 0xffe2 APP2  |    4094 | MPF_II*_.___.__.._.___0100..._.___.___..
   22258 | 0xffdb DQT   |     132 | _.......................................
   22392 | 0xffc0 SOF0  |      17 | ....p..!_........
   22411 | 0xffc4 DHT   |     418 | __........________............_.........
   22831 | 0xffda SOS  
 6196491 | 0xffd9 EOI  
 6196976 | 0xffd8 SOI  
 6196978 | 0xffe1 APP1  |    1022 | Exif__II*_.___._i.._.___._________._..._
 6198002 | 0xffdb DQT   |     132 | _..........................!.#"!. .%)4,%
 6198136 | 0xffc0 SOF0  |      17 | .......!_........
 6198155 | 0xffc4 DHT   |     418 | __........________............_.........
 6198575 | 0xffda SOS  
 6234770 | 0xffd9 EOI  
 6234864 | 0xffd8 SOI  
 6234866 | 0xffe1 APP1  |    1022 | Exif__II*_.___._i.._.___._________._..._
 6235890 | 0xffdb DQT   |     132 | _.......................................
 6236024 | 0xffc0 SOF0  |      17 | ..8.T..!_........
 6236043 | 0xffc4 DHT   |     418 | __........________............_.........
 6236463 | 0xffda SOS  
 6757985 | 0xffd9 EOI  
END: files/Stonehenge.jpg
989 rmills@rmillsm1:~/gnu/exiv2/team/book $
```

**Last word:**  The reason the program in the book is tvisitor is to explain Andreas' TiffVisitor code.  There's 10,000 lines of recursive code and that doesn't include the image handlers (jpgimage.cpp and so on).  There's 20 formats of about 500 lines.  20,000 lines of magic.  By putting everything into a single file of 4000 lines of code, the aim is to reveal how the metadata engine achieves its amazing results:

```bash
998 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ wc src/tiff* 
    1941    5863   68487 src/tiffcomposite_int.cpp
    1670    7190   67564 src/tiffcomposite_int.hpp
     118     428    3876 src/tifffwd_int.hpp
     360    1144   12812 src/tiffimage.cpp
    2123    9944  102668 src/tiffimage_int.cpp
     490    2198   19173 src/tiffimage_int.hpp
    1751    5469   60730 src/tiffvisitor_int.cpp
     736    3637   31560 src/tiffvisitor_int.hpp
    9189   35873  366870 total
999 rmills@rmillsm1:~/gnu/github/exiv2/0.27-maintenance $ wc src/tiff* 
```